File: libraries/maui-iap/src/OpenIap.Maui/WebhookClient.cs
#nullable enable

using System;
using System.Collections.Generic;
using System.IO;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using Hyo.OpenIap;

namespace Hyo.OpenIap.Maui;

public sealed record WebhookListenerOptions
{
    public required string ApiKey { get; init; }
    public string? BaseUrl { get; init; }
    public required Action<WebhookEvent> OnEvent { get; init; }
    public Action<WebhookListenerError>? OnError { get; init; }
    public HttpClient? HttpClient { get; init; }
}

public interface WebhookListener : IDisposable
{
    void Close();
}

public sealed record WebhookListenerError
{
    public required string Code { get; init; }
    public required string Message { get; init; }
    public Exception? Cause { get; init; }
}

public enum ParsedWebhookEventKind
{
    Ok,
    Skip,
    Error
}

public sealed record ParsedWebhookEventResult
{
    public required ParsedWebhookEventKind Kind { get; init; }
    public WebhookEvent? Event { get; init; }
    public string? Reason { get; init; }
    public string? Message { get; init; }
}

public static class WebhookClient
{
    private const string DefaultBaseUrl = "https://kit.openiap.dev";

    public static readonly IReadOnlyList<WebhookEventType> WebhookEventTypes =
        new[]
        {
            WebhookEventType.SubscriptionStarted,
            WebhookEventType.SubscriptionRenewed,
            WebhookEventType.SubscriptionExpired,
            WebhookEventType.SubscriptionInGracePeriod,
            WebhookEventType.SubscriptionInBillingRetry,
            WebhookEventType.SubscriptionRecovered,
            WebhookEventType.SubscriptionCanceled,
            WebhookEventType.SubscriptionUncanceled,
            WebhookEventType.SubscriptionRevoked,
            WebhookEventType.SubscriptionPriceChange,
            WebhookEventType.SubscriptionProductChanged,
            WebhookEventType.SubscriptionPaused,
            WebhookEventType.SubscriptionResumed,
            WebhookEventType.PurchaseRefunded,
            WebhookEventType.PurchaseConsumptionRequest,
            WebhookEventType.TestNotification,
        };

    public static WebhookListener ConnectWebhookStream(WebhookListenerOptions options)
    {
        if (string.IsNullOrWhiteSpace(options.ApiKey))
        {
            throw new ArgumentException("connectWebhookStream requires a non-empty apiKey.", nameof(options));
        }

        if (options.OnEvent is null)
        {
            throw new ArgumentException("connectWebhookStream requires an onEvent callback.", nameof(options));
        }

        return new HttpWebhookListener(options);
    }

    public static ParsedWebhookEventResult ParseWebhookEventData(string raw)
    {
        if (string.IsNullOrEmpty(raw))
        {
            return new ParsedWebhookEventResult
            {
                Kind = ParsedWebhookEventKind.Skip,
                Reason = "heartbeat",
            };
        }

        WebhookEvent? webhookEvent;
        try
        {
            using var doc = JsonDocument.Parse(raw);
            var root = doc.RootElement;
            if (!root.TryGetProperty("type", out var typeProp) ||
                typeProp.ValueKind != JsonValueKind.String)
            {
                return new ParsedWebhookEventResult
                {
                    Kind = ParsedWebhookEventKind.Skip,
                    Reason = "stream-control",
                };
            }

            if (!HasString(root, "id") || !HasNumber(root, "occurredAt") || !HasNumber(root, "receivedAt"))
            {
                return new ParsedWebhookEventResult
                {
                    Kind = ParsedWebhookEventKind.Error,
                    Message = "WebhookEvent missing required fields (id/occurredAt/receivedAt)",
                };
            }

            webhookEvent = JsonSerializer.Deserialize<WebhookEvent>(raw, JsonOptions.Default);
        }
        catch (JsonException ex)
        {
            return new ParsedWebhookEventResult
            {
                Kind = ParsedWebhookEventKind.Error,
                Message = $"Failed to parse SSE payload: {ex.Message}",
            };
        }

        if (webhookEvent is null)
        {
            return new ParsedWebhookEventResult
            {
                Kind = ParsedWebhookEventKind.Error,
                Message = "WebhookEvent missing required fields (id/occurredAt/receivedAt)",
            };
        }

        if (webhookEvent.Type != WebhookEventType.TestNotification &&
            string.IsNullOrEmpty(webhookEvent.PurchaseToken))
        {
            return new ParsedWebhookEventResult
            {
                Kind = ParsedWebhookEventKind.Error,
                Message = "WebhookEvent missing required field purchaseToken",
            };
        }

        return new ParsedWebhookEventResult
        {
            Kind = ParsedWebhookEventKind.Ok,
            Event = webhookEvent,
        };
    }

    private static bool HasString(JsonElement element, string propertyName)
        => element.TryGetProperty(propertyName, out var property) &&
           property.ValueKind == JsonValueKind.String;

    private static bool HasNumber(JsonElement element, string propertyName)
        => element.TryGetProperty(propertyName, out var property) &&
           property.ValueKind == JsonValueKind.Number;

    private sealed class HttpWebhookListener : WebhookListener
    {
        private readonly WebhookListenerOptions _options;
        private readonly HttpClient _http;
        private readonly bool _ownsHttpClient;
        private readonly CancellationTokenSource _cts = new();
        private readonly HashSet<string> _seenIds = new(StringComparer.Ordinal);
        private readonly Queue<string> _seenOrder = new();
        public HttpWebhookListener(WebhookListenerOptions options)
        {
            _options = options;
            _ownsHttpClient = options.HttpClient is null;
            _http = options.HttpClient ?? new HttpClient { Timeout = Timeout.InfiniteTimeSpan };
            _ = Task.Run(RunAsync);
        }

        public void Close()
        {
            if (!_cts.IsCancellationRequested)
            {
                _cts.Cancel();
            }
        }

        public void Dispose()
        {
            Close();
            _cts.Dispose();
            if (_ownsHttpClient)
            {
                _http.Dispose();
            }
        }

        private async Task RunAsync()
        {
            while (!_cts.Token.IsCancellationRequested)
            {
                try
                {
                    await ConnectOnceAsync(_cts.Token).ConfigureAwait(false);
                }
                catch (OperationCanceledException) when (_cts.Token.IsCancellationRequested)
                {
                    return;
                }
                catch (Exception ex)
                {
                    if (_cts.Token.IsCancellationRequested)
                    {
                        return;
                    }

                    _options.OnError?.Invoke(new WebhookListenerError
                    {
                        Code = "TRANSPORT_ERROR",
                        Message = "SSE transport error (auto-reconnecting)",
                        Cause = ex,
                    });
                }

                try
                {
                    await Task.Delay(TimeSpan.FromSeconds(1), _cts.Token).ConfigureAwait(false);
                }
                catch (OperationCanceledException)
                {
                    return;
                }
            }
        }

        private async Task ConnectOnceAsync(CancellationToken cancellationToken)
        {
            var baseUrl = KitApiClient.TrimTrailingSlash(_options.BaseUrl ?? DefaultBaseUrl);
            var url = $"{baseUrl}/v1/webhooks/stream/{Uri.EscapeDataString(_options.ApiKey)}";
            using var request = new HttpRequestMessage(HttpMethod.Get, url);
            request.Headers.Accept.ParseAdd("text/event-stream");

            using var response = await _http.SendAsync(
                request,
                HttpCompletionOption.ResponseHeadersRead,
                cancellationToken).ConfigureAwait(false);
            response.EnsureSuccessStatusCode();

            await using var stream = await response.Content.ReadAsStreamAsync(cancellationToken).ConfigureAwait(false);
            using var reader = new StreamReader(stream, Encoding.UTF8);
            var data = new StringBuilder();

            while (!cancellationToken.IsCancellationRequested)
            {
                var line = await reader.ReadLineAsync(cancellationToken).ConfigureAwait(false);
                if (line is null)
                {
                    return;
                }

                if (line.Length == 0)
                {
                    Dispatch(data.ToString());
                    data.Clear();
                    continue;
                }

                if (line.StartsWith("data:", StringComparison.Ordinal))
                {
                    if (data.Length > 0)
                    {
                        data.Append('\n');
                    }

                    data.Append(line[5..].TrimStart());
                }
            }
        }

        private void Dispatch(string raw)
        {
            var parsed = ParseWebhookEventData(raw);
            if (parsed.Kind == ParsedWebhookEventKind.Skip)
            {
                return;
            }

            if (parsed.Kind == ParsedWebhookEventKind.Error)
            {
                _options.OnError?.Invoke(new WebhookListenerError
                {
                    Code = "PARSE_ERROR",
                    Message = parsed.Message ?? "Failed to parse webhook event",
                });
                return;
            }

            if (parsed.Event is null || MarkSeen(parsed.Event.Id))
            {
                return;
            }

            _options.OnEvent(parsed.Event);
        }

        private bool MarkSeen(string id)
        {
            if (_seenIds.Contains(id))
            {
                return true;
            }

            _seenIds.Add(id);
            _seenOrder.Enqueue(id);
            if (_seenOrder.Count > 1024)
            {
                var evicted = _seenOrder.Dequeue();
                _seenIds.Remove(evicted);
            }

            return false;
        }
    }
}

File: libraries/expo-iap/src/webhook-client.ts
// Trusted-process transport for the openiap kit SSE stream. The
// administrative route is `GET /v1/webhooks/stream` with a secret Bearer key.
// Never embed the secret in a shipped app.
//
// The wire format is documented in `packages/kit/server/api/v1/webhooks.ts`
// and matches the GraphQL `WebhookEvent` shape from `webhook.graphql`.
//
// Parser logic is split out from the connection so it can be unit-
// tested without a live server. See `webhook-client.test.ts`.

export type WebhookEventType =
  | "SubscriptionStarted"
  | "SubscriptionRenewed"
  | "SubscriptionExpired"
  | "SubscriptionInGracePeriod"
  | "SubscriptionInBillingRetry"
  | "SubscriptionRecovered"
  | "SubscriptionCanceled"
  | "SubscriptionUncanceled"
  | "SubscriptionRevoked"
  | "SubscriptionPriceChange"
  | "SubscriptionProductChanged"
  | "SubscriptionPaused"
  | "SubscriptionResumed"
  | "PurchaseRefunded"
  | "PurchaseConsumptionRequest"
  | "TestNotification";

export const WEBHOOK_EVENT_TYPES = [
  "SubscriptionStarted",
  "SubscriptionRenewed",
  "SubscriptionExpired",
  "SubscriptionInGracePeriod",
  "SubscriptionInBillingRetry",
  "SubscriptionRecovered",
  "SubscriptionCanceled",
  "SubscriptionUncanceled",
  "SubscriptionRevoked",
  "SubscriptionPriceChange",
  "SubscriptionProductChanged",
  "SubscriptionPaused",
  "SubscriptionResumed",
  "PurchaseRefunded",
  "PurchaseConsumptionRequest",
  "TestNotification",
] as const satisfies readonly WebhookEventType[];

export type WebhookEventPayload = {
  id: string;
  type: WebhookEventType;
  source: string;
  platform: "IOS" | "Android";
  environment: "Production" | "Sandbox" | "Xcode";
  projectId: string;
  occurredAt: number;
  receivedAt: number;
  // Optional because TestNotification frames carry no transaction;
  // every other event type populates this.
  purchaseToken?: string;
  productId?: string;
  subscriptionState?: string;
  expiresAt?: number;
  renewsAt?: number;
  cancellationReason?: string;
  currency?: string;
  priceAmountMicros?: number;
  rawSignedPayload?: string;
};

export type WebhookListenerOptions = {
  /**
   * Secret admin key sent in the Authorization header. Never put this value in
   * a shipped app.
   */
  apiKey: string;
  /**
   * Override the kit base URL. Defaults to https://kit.openiap.dev.
   * In tests, point this at a local server.
   */
  baseUrl?: string;
  /** Called on every successfully-parsed webhook event. */
  onEvent: (event: WebhookEventPayload) => void;
  /**
   * Called on transport errors. The connection auto-reconnects
   * unconditionally; this callback exists for telemetry / surfacing
   * to the host UI.
   */
  onError?: (error: WebhookListenerError) => void;
  /**
   * Optional injection of an EventSource constructor. Lets RN /
   * Expo plug in `react-native-event-source` when running on a JS
   * runtime that lacks the global, or vitest plug in a stub.
   */
  eventSourceFactory?: (
    url: string,
    headers: Record<string, string>,
  ) => WebhookEventStream;
};

export interface WebhookEventStream {
  close(): void;
  onmessage: ((event: { data: string; lastEventId?: string }) => void) | null;
  onerror: ((error: unknown) => void) | null;
  addEventListener?: (
    type: string,
    listener: (event: { data: string; lastEventId?: string }) => void,
  ) => void;
}

export type WebhookListener = {
  /** Tear down the connection and stop receiving events. */
  close(): void;
};

export type WebhookListenerError = {
  code:
    "TRANSPORT_ERROR" | "PARSE_ERROR" | "MALFORMED_EVENT" | "NO_EVENTSOURCE";
  message: string;
  cause?: unknown;
};

const DEFAULT_BASE_URL = "https://kit.openiap.dev";

/**
 * Trusted-process transport for the project-wide IAPKit stream.
 *
 * Hosted IAPKit now requires a secret admin key for this stream. Never call
 * this helper from a shipped app or pass it a publishable key. The secret is
 * sent in the Authorization header so it does not enter URLs or proxy logs.
 */
export function connectWebhookStream(
  options: WebhookListenerOptions,
): WebhookListener {
  const baseUrl = options.baseUrl ?? DEFAULT_BASE_URL;
  const url = `${trimTrailingSlash(baseUrl)}/v1/webhooks/stream`;

  const factory = options.eventSourceFactory ?? defaultEventSourceFactory;
  let stream: WebhookEventStream;
  try {
    stream = factory(url, {
      Authorization: `Bearer ${options.apiKey}`,
    });
  } catch (error) {
    options.onError?.({
      code: "NO_EVENTSOURCE",
      message:
        error instanceof Error
          ? error.message
          : "EventSource constructor unavailable in this runtime",
      cause: error,
    });
    return { close: () => {} };
  }

  const seenIds = new Set<string>();
  const seenOrder: string[] = [];
  const markSeen = (id: string): boolean => {
    if (seenIds.has(id)) {
      return true;
    }
    seenIds.add(id);
    seenOrder.push(id);
    if (seenOrder.length > 1024) {
      const evicted = seenOrder.shift();
      if (evicted !== undefined) {
        seenIds.delete(evicted);
      }
    }
    return false;
  };

  const handleData = (raw: string) => {
    const parsed = parseWebhookEventData(raw);
    if (parsed.kind === "error") {
      options.onError?.({
        code: "PARSE_ERROR",
        message: parsed.message,
      });
      return;
    }
    if (parsed.kind === "skip") {
      return;
    }
    if (markSeen(parsed.event.id)) {
      return;
    }
    options.onEvent(parsed.event);
  };

  if (typeof stream.addEventListener === "function") {
    stream.addEventListener("message", (event) => handleData(event.data));
    // WHATWG EventSource dispatches frames with `event: Foo` only to
    // listeners registered for `Foo`, not to `message` / `onmessage`.
    // Kit emits webhook frames as typed SSE events, so subscribe to
    // every known webhook type and keep `message` for older servers or
    // polyfills that collapse typed frames into the generic channel.
    for (const eventType of WEBHOOK_EVENT_TYPES) {
      stream.addEventListener(eventType, (event) => handleData(event.data));
    }
  } else {
    stream.onmessage = (event) => handleData(event.data);
  }

  stream.onerror = (error) => {
    options.onError?.({
      code: "TRANSPORT_ERROR",
      message: "SSE transport error (auto-reconnecting)",
      cause: error,
    });
  };

  return {
    close: () => {
      try {
        stream.close();
      } catch {
        // Closing an already-closed EventSource is a no-op in browsers
        // but throws in some polyfills.
      }
    },
  };
}

// ---------------------------------------------------------------------------
// Pure helpers (exported for testing).
// ---------------------------------------------------------------------------

export type ParsedEventResult =
  | { kind: "ok"; event: WebhookEventPayload }
  | { kind: "skip"; reason: "heartbeat" | "stream-control" }
  | { kind: "error"; message: string };

export function parseWebhookEventData(raw: string): ParsedEventResult {
  if (!raw) {
    return { kind: "skip", reason: "heartbeat" };
  }

  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch (error) {
    return {
      kind: "error",
      message:
        error instanceof Error
          ? `Failed to parse SSE payload: ${error.message}`
          : "Failed to parse SSE payload",
    };
  }

  if (
    typeof parsed !== "object" ||
    parsed === null ||
    !("type" in parsed) ||
    typeof (parsed as Record<string, unknown>).type !== "string"
  ) {
    // Stream-control messages (the `ready`/`stream-error` envelopes
    // emitted by the kit server) have no `type` and are surfaced as
    // skips so consumers don't see them as events.
    return { kind: "skip", reason: "stream-control" };
  }

  const event = parsed as WebhookEventPayload;

  if (
    typeof event.id !== "string" ||
    typeof event.occurredAt !== "number" ||
    typeof event.receivedAt !== "number"
  ) {
    return {
      kind: "error",
      message: `WebhookEvent missing required fields (id/occurredAt/receivedAt)`,
    };
  }
  // purchaseToken is required for every event type *except*
  // TestNotification — Apple ASN v2 / Google RTDN test payloads
  // carry no transaction. Hard-rejecting here would surface valid
  // test webhooks as MALFORMED_EVENT and never reach listeners.
  if (
    event.type !== "TestNotification" &&
    typeof event.purchaseToken !== "string"
  ) {
    return {
      kind: "error",
      message: `WebhookEvent missing required field purchaseToken`,
    };
  }

  return { kind: "ok", event };
}

function trimTrailingSlash(url: string): string {
  return url.endsWith("/") ? url.slice(0, -1) : url;
}

function defaultEventSourceFactory(
  _url: string,
  _headers: Record<string, string>,
): WebhookEventStream {
  // WHATWG EventSource cannot attach an Authorization header. Fail closed
  // instead of falling back to a secret-in-URL route or silently opening an
  // unauthenticated stream.
  throw new Error(
    "Pass an eventSourceFactory that supports Authorization headers.",
  );
}
